Minimize System.Version formatting buffer's length

Chasmical opened the pull request:
Change the length of the char buffer used in `Version.ToString()` from 47 to 43 chars (4 non-negative `Int32`s + 3 periods). `2147483647.2147483647.2147483647.2147483647`. Even if a version has invalid internal state (any component less than -1), the `TryFormat` method just casts them to `UInt32` anyway.

Chasmical committed:
Minimize System.Version formatting buffer's length

Change the length of the char buffer used in Version.ToString()
from 47 to 43 chars (4 non-negative Int32s + 3 periods). Even if
a version has invalid internal state (any component less than -1),
the TryFormat method just casts them to UInt32 anyway.

Chasmical committed:
Fix usage of UInt32NumberBufferLength

EgorBo commented:
To be fair, this doesn't reduce the actual stack allocation size - JIT is going to round up both to 8 (or 16?) anyway, e.g. see [sharplab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4am2uSRNSDIQwDeNBmYanzls813EUDACqqMAfQCy2XlwAUASmtjQPMGSC5cDAZvSIBJHgBmUgA5DnxgGQAhDgAzbJkAGRguAHMMAAsGAF4GckYAahqNahCQ2TFsLgAeMDLsKAA+BmAcvKgqhgjsMABrbElJCDBQ3qgAbR8HACoGOIxElLTMkYKi0rK/Bgb4gF0mlvMRcNSYH2HcmT878wBfQMDbJgOZwRVwABTY/kCJma91CEHCkWiOwSyVS6SgWXeUEKJXK41ql0awXMbQ63RWgzeo3GkxmcwWSx6fXWWwYPl2+zRRyxOLODDgNT8Fyut2JZkeuGer2OUE+gV+1ECq3cMHKEAAJjFxJIfCq1ZrtQB5MR8eG4FhJCBxSTebzFPzXf46QEGM1S0ldJkDBiuC5GBgKmjfIA)

Chasmical committed:
Replace Int32NumberBufferLength with const

Number.Int32NumberBufferLength is only used in System.Number class and
here. The extra 1 to length is designated for the terminating nulls.
It would make more sense not to use it here.

Chasmical commented:
I see. 😄 Then, I guess it could still improve readability of the decompiled code. It was confusing to see 47 characters allocated when the actual maximum length is 43.

On another note, looks like `Number.Int32NumberBufferLength` isn't used anywhere else, besides the `Number` class and here, and the extra 1 to length is actually for the terminating nulls. Would make more sense not to use it here, in case `Number.NumberBuffer` changes.

https://github.com/dotnet/runtime/blob/153a94b88d77cfe133682e05191bc41914af6b21/src/libraries/Common/src/System/Number.NumberBuffer.cs#L13-L16

stephentoub commented:
Thank you, but I'd rather keep it how it is today. I think it's more maintainable as it is currently and this doesn't actually improve perf.
